+= overloads for sparse subdomains

e-kayrakli opened the pull request:
This pull request adds some += overloads for non-distributed sparse subdomains.

Following lines call bulkAdd that inserts indices _hopefully_ efficiently. Some basic performance tests(not included) showed couple orders of magnitude speedup compared to adding indices one by one, which is not surprising.

```
SD += array_of_indices; //not necessarily sorted, can have duplicates
SD += other_dense_dom;
SD += other_sparse_dom;
```

In case of adding array of indices, bulkAdd can be called directly to avoid sorting and duplicate checking. But it's user's responsibility to make sure that the indices are actually sorted and have no duplicates.

```
SD.bulkAdd(sorted_nodup_indices, true, true);
```

PR also adds some tests for correctness.

Few ideas for future steps on this:

- There _might be_ room for some optimization for adding sparse subdomains, especially when their layout is the same. This PR iterates over the indices of RHS domain to create an array of indexes.
- += for `SparseBlockDom`

e-kayrakli commented:
Assigned @mppf as we per our previous discussions.

Also, it passed the correctness tests with default environment on chapcs.

mppf commented on modules/internal/ChapelArray.chpl:
```diff
@@ -1030,6 +1030,16 @@ module ChapelArray {
     }
 
+    proc bulkAdd(inds: [] _value.idxType, sorted=false,
+        noDuplicate=false) where isSparseDom(this) && _value.rank==1 {
```
I like sorted=false but I'd consider renaming 'noDuplicate' to 'noDuplicates' or 'unique'

mppf commented on modules/internal/ChapelArray.chpl:
```diff
@@ -1473,6 +1483,34 @@ module ChapelArray {
   }  // record _domain
 
+  proc +=(ref sd: domain, inds: [] sd.idxType) 
+    where isSparseDom(sd) && sd.rank == 1{
```
Could you add a space before the {   ?

mppf commented on modules/internal/ChapelArray.chpl:
```diff
@@ -1030,6 +1030,16 @@ module ChapelArray {
+    }
+
+    proc bulkAdd(inds: [] _value.rank*_value.idxType, sorted=false,
+        noDuplicate=false) where isSparseDom(this) && _value.rank>1{
```
Here too I'd appreciate a space before {

mppf commented on modules/internal/ChapelArray.chpl:
```diff
@@ -1473,6 +1483,34 @@ module ChapelArray {
+    const indCount = d._value.dsiNumIndices;
+    const arr: [{0..#indCount}] _idxType;
+  
+    //this could be a parallel loop. but ranks don't match -- doesn't compile
```
I think we need to think about rank-change zippered iteration... but here I am wondering why bulkAdd cares at all about the rank? Is it because we might sort it, so want it to be 1-D ?

e-kayrakli replied:
I somehow couldn't find a universal way of getting all indices in a domain(including dense and sparse), and a zippered forall over 1-D array and 2-D domain didn't work obviously.

mppf replied:
I don't think you need to change this PR, but put the issue on your mental list to think about.

mppf commented on modules/internal/DefaultSparse.chpl:
```diff
@@ -406,6 +513,21 @@ module DefaultSparse {
       return irv;
     }
 
+    proc sparseBulkShiftArray(shiftMap, oldnnz){
```
It's probably obvious to you what sparseBulkShiftArray is doing and why, but I don't understand immediately. Could you add a comment to it describing what it does here? I think it would be also nice to have a comment describing sparseShiftArray. I don't already know what either of these do...

mppf commented on modules/internal/ChapelArray.chpl:
```diff
@@ -1473,6 +1483,34 @@ module ChapelArray {
+    where isSparseDom(sd) && d.rank==sd.rank && sd.idxType==d.idxType {
+
+    type _idxType = if sd.rank==1 then int else sd.rank*int;
+    const indCount = d._value.dsiNumIndices;
```
I don't know of any reason not to use d.size here.

mppf replied:
or d.numIndices for that matter.

mppf commented on modules/internal/DefaultSparse.chpl:
```diff
@@ -235,7 +235,114 @@ module DefaultSparse {
+      var indivInsertPts: [{0..#numAdded}] int;
+      var actualInsertPts: [{0..#numAdded}] int; //where to put in newdom
+
+      if !sorted then QuickSort(inds);
```
It seems surpsising to me that the passed array will be modified. E.g. if I do

      someDomain += myArray;

it's not immediately obvious when looking that myArray will be modified.
It could be documented well, but I think it's probably better to sort the
array in temporary space. Another interesting possibility would be to
allow bulkAdd to use an argument indicating if inds can be written to.
But, if += is not really useful if the array is copied and then sorted...
then we could talk more about how to make this clearer.

mppf commented on modules/internal/DefaultSparse.chpl:
```diff
@@ -235,7 +235,114 @@ module DefaultSparse {
+        const indsEnd = inds.domain.high;
+        var lastInd = inds[indsStart];
+        for i in indsStart+1..indsEnd {
+          if inds[i] == lastInd then halt("There are duplicates"); 
```
How about ?

bulkAdd: There are duplicates, call the function with unique=false

mppf commented on modules/internal/DefaultSparse.chpl:
```diff
@@ -235,7 +235,114 @@ module DefaultSparse {
+      }
+
+      for (i,p) in zip(inds, indivInsertPts) {
+        if p != -1 { //don't do anything if it's duplicate
```
Isn't indivInsertPtrs al 0s if noDuplicate=true at this point?

mppf replied:
Is there any benefit to parallelizing this loop?

mppf commented on modules/internal/DefaultSparse.chpl:
```diff
@@ -235,7 +235,114 @@ module DefaultSparse {
+      //shift insert points for bulk addition
+      //previous indexes that are added will cause a shift in the next indexes
+      var actualAddCnt = 0;
+      for (ip, ap) in zip(indivInsertPts, actualInsertPts) {
```
FYI I think this could be implemented with scan, but it might be better to do it this way for now.

mppf commented on modules/internal/DefaultSparse.chpl:
```diff
@@ -235,7 +235,114 @@ module DefaultSparse {
+        //make sure lastInd != inds[inds.domain.low]
+        var lastInd = inds[inds.domain.low] + 1; 
+        for (i, p) in zip(inds, indivInsertPts)  {
+          if i == lastInd then p = -1;
```
Use of -1 in a bunch of places here might be better handled by parentDom.low - 1 or something like that. -1 wouldn't work with unsigned index types. Of course, it wouldn't surprise me if there are other problems with unsigned index types for sparse domains... so it might be moot.

e-kayrakli replied:
Actually -1 there is a marker that is not relevant to parentDom but nnzDom. And nnzDom is hardcoded to be 1-based. If I use nnzDom.low-1, the caveat is that it will be equal to 0 which is the default initial value of the array, and to avoid that there has to be an initialization to another value, which brings us to where we began. Am I missing something?

mppf replied:
Find with me then. Would it work if the "zero value" was actually -1 ?

e-kayrakli replied:
It seems like it. Do you think it worth changing? I can also define a local constant like a `NOOP_MARKER` that is -1, if you think it would clear the usage a bit more?

mppf commented on modules/internal/DefaultSparse.chpl:
```diff
@@ -235,7 +235,114 @@ module DefaultSparse {
+          if inds[i] == lastInd then halt("There are duplicates"); 
+        }
+
+        //couldn't use boundsCheck because of 1D indices -- they are not 1tuple
```
Maybe just make boundsCheck accept both forms? Or convert it here?

mppf commented on modules/layouts/LayoutCSR.chpl:
```diff
@@ -287,6 +288,133 @@ class CSRDom: BaseSparseDom {
+        actualAddCnt += 1;
+      }
+      else ap = ip;
+    }
```
Up to this point, the code is almost identical to the DefaultSparse version. Could you make it into some sort of helper routine?

mppf commented on modules/internal/ChapelArray.chpl:
```diff
@@ -1030,6 +1030,16 @@ module ChapelArray {
+
+    proc bulkAdd(inds: [] _value.rank*_value.idxType, sorted=false,
+        noDuplicate=false) where isSparseDom(this) && _value.rank>1{
+      _value.bulkAdd(inds, sorted, noDuplicate);
```
I'd expect it to be domain.bulkAdd but domain._value.dsiBulkAdd.

mppf commented:
@e-kayrakli - cool! Overall, I have three main points of feedback:

1) I think the DefaultSparse/CSR Doms classes should implement dsiBulkAdd, not bulkAdd. But, if Brad thinks there's no reason to prefix these methods with dsi, that's fine too.
2) I'd like to have a better name for noDuplicate. It might seem trivial, but since we have pass-by-name, these argument names are part of the API. I like 'isUnique' , 'uniqued', or 'noDuplicates' but it's probably worth asking somebody else too for 10 seconds... Another alternative is to have a different bulkAdd function that applies when the data is sorted.
3) I'm concerned about the += sorting/uniquing data in place. For 'bulkAdd' or related functions, I think it could be an issue for documentation.

Thanks!

e-kayrakli commented:
Last batch of commits fixes several minor issues that Michael brought up. I replied some of his comments which are not directly visible due to outdated diff.

To reiterate open questions:

- Should the name be `bulkAdd` or `dsiBulkAdd`? I don't know if there are any implications of prefixing a function name with "dsi", other than convention. @bradcray may have something to say on that.

- I changed `noDuplicate` flag to `isUnique` and changed `sorted` to `isSorted` for consistency. But I guess it is still open for discussion.

- Big question: Do we sort/unique in-place? I totally agree that when user writes `SD+=my_index_array` and finds `my_index_array` totally ruined by that is unacceptable. I am currently thinking about an additional flag like `preserveInds` that defaults to false. It will have memory and performance implications but += will be safe. Users who know what they are doing can still call `SD.bulkAdd(my_index_array, true, true, true)`. We can even provide a wrapper like `fastBulkAdd`, that passes `true, true, true` to `bulkAdd` to keep things clean.

e-kayrakli commented:
How about a compile or runtime constant to make += call `bulkAdd(X, true, true, true)`?

mppf commented:
I think I'd prefer for += to always use a preserveInds=true version (ie the version that does not modify the passed array) and then rely on users to call SD.bulkAdd if they want the other behavior.

bradcray commented:
Typically the 'dsi' prefix is for routines that are expected to be supported by the domain map classes (stands for "domain map standard interface" and could be considered a naming convention in the place of a real language-level concept for interfaces).  We also sometimes use the 'doi' prefix for routines that are optional and not required ("domain map optional interface").  So the standard thing to do here would typically be to make the implementations of these routines in the specific distributions/layouts named with 'dsi' and the top-level routines exposed by _domain and _array be something more user-palatable (e.g., bulkAdd).  New dsi* routines should also ideally be documented in doc/release/technotes/dsi.rst.

e-kayrakli commented:
Are there any other concerns that I haven't addressed? Is it ready to merge?

mppf commented on modules/internal/ChapelArray.chpl:
```diff
@@ -1473,6 +1485,98 @@ module ChapelArray {
     }
   }  // record _domain
 
+  proc __getActualInsertPts(d, inds, 
```
I'm not so sure this function belongs in ChapelArray. At least it needs a comment indicating it's a helper method for DefaultSparse/CSR.

e-kayrakli replied:
I agree. This issue seems to be a slippery slope. As far as I can see ChapelArray.chpl has some functions that should be moved to other files/classes. eg. BaseSparseDom in ChapelDistribution.chpl is the most suitable class for this function to be IMO. But for consistency I'd like to keep it in ChapelArray(for now at least). I can come back to this at a larger refactoring context.

mppf commented:
Other than adding that comment, it's fine with me to merge once you are satisfied with testing. I'd check some tests in a GASNet configuration in addition to the standard testing.

e-kayrakli commented:
Passed whole suite with GASNet configuration. Merging.
